larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,624 @@
+
+    TwoDSlidingFitResultMap microSlidingFitResultMap;
+    TwoDSlidingFitResultMap macroSlidingFitResultMap;
+    this->InitialiseSlidingFitResultMaps(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+    // Find longest length association and conitnue merging until no further merges can be made
+    bool mergeMade(false);
+    do
+    {
+        ClusterAssociation clusterAssociation;
+        if(!this->FindBestClusterAssociation(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap, clusterAssociation))
+            break;
+                        
+        CaloHitToParentClusterMap caloHitToParentClusterMap;
+        CaloHitVector extrapolatedCaloHitVector;
+        this->GetExtrapolatedCaloHits(clusterAssociation, pClusterList, extrapolatedCaloHitVector, caloHitToParentClusterMap);
+
+        if (extrapolatedCaloHitVector.empty())
+            break;
+        
+        if (!this->IsTrackContinuous(clusterAssociation, extrapolatedCaloHitVector))
+            break;
+        
+        this->RefineTracks(clusterAssociation, extrapolatedCaloHitVector, microSlidingFitResultMap);
+        
+        this->AddHitsToCluster(clusterAssociation, caloHitToParentClusterMap, extrapolatedCaloHitVector, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+        this->UpdateSlidingFitResultMap(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+        
+        mergeMade = true;
+    } while (mergeMade);
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::SelectCleanClusters(const ClusterList *pClusterList, ClusterVector &clusterVector) const
+{
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+        if (pCluster->GetNCaloHits() < m_minCaloHits)
+            continue;
+
+        clusterVector.push_back(pCluster);
+    }
+    
+    std::sort(clusterVector.begin(), clusterVector.end(), LArClusterHelper::SortByNHits);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::InitialiseSlidingFitResultMaps(const ClusterVector &clusterVector, TwoDSlidingFitResultMap &microSlidingFitResultMap,
+    TwoDSlidingFitResultMap &macroSlidingFitResultMap) const
+{
+    const float slidingFitPitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+    
+    for (const Cluster *const pCluster : clusterVector)
+    {
+        try
+        {
+            (void) microSlidingFitResultMap.insert(TwoDSlidingFitResultMap::value_type(pCluster, TwoDSlidingFitResult(pCluster, m_slidingFitWindow, slidingFitPitch)));
+            (void) macroSlidingFitResultMap.insert(TwoDSlidingFitResultMap::value_type(pCluster, TwoDSlidingFitResult(pCluster, 1000000, slidingFitPitch)));
```

Comment:
So when I originally done this I found serious performance changes and also found that the number of entries in the map falls, suggesting fits in some cases are no longer made. I wasn't sure if there was a limit on the sliding fit window after which the fit no longer made sense. But I can have another look at this.

Reply:
I can see why the computational performance would change, as this upper limit appears to be used as a condition for loop termination in the fit, so uint::max implies O(10^9) iteration loops, so that would seem to incentivise the use of the smallest number you can get away with. It's interesting that the fit would have different outcomes for larger values, given that 10^6 seems like a large enough number for any such issue to appear in practice. Certainly worth investigating a sensible number here and assigning it to some named variable/const to make it clear what the value is being used for.